src/indieweb_utils/webmentions/discovery.py

```diff
@@ -71,29 +74,51 @@ def _discover_endpoints(url, headers_to_find):
     :return: The discovered endpoints.
     :rtype: dict[str, str]
     """
-    response = {}
+    response: Dict[str, str] = {}
 
     endpoint_request = requests.get(url)
 
-    http_link_headers = endpoint_request.headers.get("link")
+    response.update(_find_links_in_headers(headers=endpoint_request.headers, target_headers=headers_to_find))
+    response.update(_find_links_html(body=endpoint_request.text, target_headers=headers_to_find))
+    return response
+
 
-    if http_link_headers:
-        parsed_link_headers = requests.utils.parse_header_links(http_link_headers.rstrip(">").replace(">,<", ",<"))
+def _find_links_in_headers(*, headers, target_headers: List[str]) -> Dict[str, str]:
+    """Return a dictionary { rel: url } containing the target headers."""
+    found: Dict[str, str] = {}
+    links = headers.get("link")
+    if links:
+        # [{'url': 'https://micropub.jamesg.blog/micropub', 'rel': 'micropub'} ]
+        parsed_link_headers: List[Dict[str, str]] = requests.utils.parse_header_links(links)
     else:
-        parsed_link_headers = []
+        return found
 
     for header in parsed_link_headers:
-        if header["rel"] in headers_to_find:
-            response[header["rel"]] = header["url"]
+        url = header.get("url", "")
+        rel = header.get("rel", "")
+        if _is_http_url(url) and rel in target_headers:
+            found[rel] = url
+    return found
+
 
-    soup = BeautifulSoup(endpoint_request.text, "lxml")
+def _find_links_html(*, body: str, target_headers: List[str]) -> Dict[str, str]:
+    """"""
+    soup = BeautifulSoup(body, "html.parser")
```

Comment:
Use html.parser because we're parsing html.

Reply:
Are there any particular advantages to using html.parser over lxml? I think lxml is the default with BeautifulSoup. I have read in the past that lxml is faster.

Reply:
The docs have a good [overview of the differences](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#differences-between-parsers). What I like about the html.parser over lxml or html5lib parser is that it won't try to make the document "well formed" and add `<html>` and `<body>` tags – that's bitten before when developing Tanzawa.

Reply:
Thank you for sharing! I am happy to use html.parser. I use it as a fallback on IndieWeb Search I believe and haven't run into any issues using it.